File: project/8.06_独自の例外/py/py_ex8_06_x/src/ExceptionSubclass.py
'''
Created on 2023/05/17

@author: Keroichi.T
'''
import random

# 独自の例外クラスを作成できる
class MyErrorA( Exception ):
    def __init__(self,value):
        self.value = f"MyErrorA: {value}"
    
    def __str__(self):
        return repr(self.value)

class MyErrorB( Exception ):
    def __init__(self,value):
        self.value = f"MyErrorB: {value}"
    
    def __str__(self):
        return repr(self.value)

class MyErrorM( Exception ):
    def __init__(self,value):
        self.value = f"MyErrorM: {value}"
    
    def __str__(self):
        return repr(self.value)

class MyErrorN( Exception ):
    def __init__(self,value):
        self.value = f"MyErrorN: {value}"
    
    def __str__(self):
        return repr(self.value)
    


def main():
    a()
    
def a():
    try:
        try:
            raise MyErrorM( "@ inner try" )
    
        except MyErrorM as e:
            print( f"Inner catch({e})" )
        
        b()
        raise MyErrorM( "@ outer try" )

    except MyErrorM as e:
        print( f"Outer catch({e})" )
        
    except Exception as e:
        print( f"Outer catch({e})" )

def b():
    try:
        c()
        
    except MyErrorB as e:
        print( f"catch @ b method({e})" )
        raise

def c():
    i = random.randint(1,20)
    if i % 2 == 0:
        raise MyErrorA( f"number: {i}" )
    else:
        raise MyErrorB( f"number: {i}" )

if __name__ == '__main__':
    main()

File: project/8.06_独自の例外/cpp/cpp_ex8_06_x/src/ExceptionSubclass.cpp
//============================================================================
// Name        : cpp_ex8_06_x.cpp
// Author      : keroichi.T
// Version     :
// Copyright   : Your copyright notice
// Description : Hello World in C++, Ansi-style
//============================================================================

#include <iostream>
#include <random>

using namespace std;

// 独自の例外クラスを作成
class MyErrorA: public exception {
public:
	const char *what() const noexcept(true) {
		return "MyErrorA\n";
	}
};

class MyErrorB: public exception {
public:
	const char *what() const noexcept(true) {
		return "MyErrorB\n";
	}
};

class MyErrorM: public exception {
public:
	const char *what() const noexcept(true) {
		return "MyErrorM\n";
	}
};

class MyErrorN: public exception {
public:
	const char *what() const noexcept(true) {
		return "MyErrorN\n";
	}
};

static void a() noexcept(false);
static void b() noexcept(false);
static void c() noexcept(false);

int main() {

	a();

	return 0;
}

static void a() {

	try {
		try {
			throw MyErrorM();
//			throw MyErrorN();
		}
		catch( MyErrorN &e ) {
			cout << "a: Inner catch " << e.what() << endl;
		}
		b();
		throw MyErrorM();
	}
	catch( MyErrorM &e ) {
		cout << "a: Outer catch " << e.what() << endl;
	}
	catch( std::exception &e ) {

	}
}

static void b() {

	try {
		c();
	}
	catch( MyErrorB &e ) {
		cout << "b: MyErrorB " << e.what() << endl;
	}
}

static void c() {

	std::random_device rnd;
	int r = rnd() % 20 + 1;
	cout << r << endl;
	if( r % 2 == 0 ) {
		throw MyErrorA();
	}
	else {
		throw MyErrorB();
	}

}
